storage.lua

```diff
@@ -1,47 +1,194 @@
+-- storage getter/setter
 
-function mail.getMailFile(playername)
-	local saneplayername = string.gsub(playername, "[.|/]", "")
-	return mail.maildir .. "/" .. saneplayername .. ".json"
+-- create or populate empty fields on an entry
+local function populate_entry(e)
+	e = e or {}
+	e.contacts = e.contacts or {}
+	e.inbox = e.inbox or {}
+	e.outbox = e.outbox or {}
+	e.lists = e.lists or {}
+	return e
 end
 
-function mail.getContactsFile(playername)
-	local saneplayername = string.gsub(playername, "[.|/]", "")
-	return mail.maildir .. "/contacts/" .. saneplayername .. ".json"
+function mail.get_storage_entry(playername)
+	local str = mail.storage:get_string(playername)
+	if str == "" then
+		-- new entry
+		return populate_entry()
+	else
+		-- deserialize existing entry
+		local e = minetest.parse_json(str)
+		return populate_entry(e)
+	end
 end
 
+function mail.set_storage_entry(playername, entry)
+	mail.storage:set_string(playername, minetest.write_json(entry))
+end
```

Comment:
It would probably be good to add some static prefix to player storage keys so that there's clean keys for other stuff if needed later (without prefixes while possible it'll be annoying if one day you'd like to store some other stuff).

I think just adding any static prefix here should work without any other changes, like even just `"mail/" .. playername`.

Reply:
done :+1: